ComInterfaceGenerator emits method attributed with [UnmanagedCallersOnly] and `in` `out` or `ref` parameters.

ComInterfaceGenerator generates code that produces `error CS8977: Cannot use 'ref', 'in', or 'out' in the signature of a method attributed with 'UnmanagedCallersOnly'.`

User Code:
```C#
[assembly:DisableRuntimeMarshalling]

[global::System.Runtime.InteropServices.Marshalling.GeneratedComInterfaceAttribute]
partial interface INativeAPI : IUnmanagedInterfaceType
{
    
    void Method(out S value);
}
// Try using the generated native interface
sealed class NativeAPI : IUnmanagedVirtualMethodTableProvider, INativeAPI.Native
{
    public VirtualMethodTableInfo GetVirtualMethodTableInfoForKey(System.Type type) => throw null;
}
partial interface INativeAPI
{
        static unsafe void* IUnmanagedInterfaceType.VirtualMethodTableManagedImplementation => null;
}
[NativeMarshalling(typeof(Marshaller))]
public struct S
{
#pragma warning disable CS0649 // Field is never assigned to, and will always have its default value
    public bool b;
#pragma warning restore CS0649
}

[CustomMarshaller(typeof(S), MarshalMode.ManagedToUnmanagedOut, typeof(Marshaller))]
[CustomMarshaller(typeof(S), MarshalMode.UnmanagedToManagedIn, typeof(Marshaller))]
public static class Marshaller
{
    public struct Native { }

    public static S ConvertToManaged(Native n) => default;
}
```

NativeToManagedStubs.g.cs
```C#
// <auto-generated/>
unsafe partial interface INativeAPI
{
    internal unsafe partial interface Native
    {
        // error CS8977: Cannot use 'ref', 'in', or 'out' in the signature of a method attributed with 'UnmanagedCallersOnly'.
        [System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute]
        internal static void ABI_Method(void* __this_native, out global::S value)
        {
            global::INativeAPI @this = default;
            try
            {
                // Unmarshal - Convert native data to managed data.
                @this = (global::INativeAPI)System.Runtime.InteropServices.Marshalling.UnmanagedObjectUnwrapper.GetObjectForUnmanagedWrapper<System.Runtime.InteropServices.Marshalling.ComWrappersUnwrapper>(__this_native);
                @this.Method(out value);
            }
            catch (System.Exception __exception)
            {
                System.Runtime.InteropServices.Marshalling.SwallowExceptionMarshaller.ConvertToUnmanaged(__exception);
            }
        }
    }
}
```

--- a/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
@@ -580,7 +580,7 @@ namespace Microsoft.Interop
 
             var vtableSlotAssignments = VirtualMethodPointerStubGenerator.GenerateVirtualMethodTableSlotAssignments(
                 interfaceMethods.Methods
-                    .Where(context => context.GenerationContext.Diagnostics.Length == 0 && context.UnmanagedToManagedStub.Diagnostics.Length == 0)
+                    .Where(context => context.UnmanagedToManagedStub.Diagnostics.Length == 0)
                     .Select(context => context.GenerationContext),
                 vtableLocalName);
 

--- a/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
@@ -417,6 +417,7 @@ namespace Microsoft.Interop
                         comInterfaceAndMethods.DeclaredMethods
                             .Select(m => m.UnmanagedToManagedStub)
                             .OfType<GeneratedStubCodeContext>()
+                            .Where(context => context.Diagnostics.Length == 0)
                             .Select(context => context.Stub.Node)));
         }
 
@@ -577,7 +578,11 @@ namespace Microsoft.Interop
                                         })))));
             }
 
-            var vtableSlotAssignments = VirtualMethodPointerStubGenerator.GenerateVirtualMethodTableSlotAssignments(interfaceMethodStubs, vtableLocalName);
+            var vtableSlotAssignments = VirtualMethodPointerStubGenerator.GenerateVirtualMethodTableSlotAssignments(
+                interfaceMethods.Methods
+                    .Where(context => context.GenerationContext.Diagnostics.Length == 0 && context.UnmanagedToManagedStub.Diagnostics.Length == 0)
+                    .Select(context => context.GenerationContext),
+                vtableLocalName);
 
             return ImplementationInterfaceTemplate
                 .AddMembers(

--- a/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
@@ -437,7 +437,6 @@ namespace Microsoft.Interop
 
             const string vtableLocalName = "vtable";
             var interfaceType = interfaceMethods.Interface.Info.Type;
-            var interfaceMethodStubs = interfaceMethods.DeclaredMethods.Select(m => m.GenerationContext);
 
             // void** vtable = (void**)RuntimeHelpers.AllocateTypeAssociatedMemory(<interfaceType>, sizeof(void*) * <max(vtableIndex) + 1>);
             var vtableDeclarationStatement =
@@ -579,7 +578,7 @@ namespace Microsoft.Interop
             }
 
             var vtableSlotAssignments = VirtualMethodPointerStubGenerator.GenerateVirtualMethodTableSlotAssignments(
-                interfaceMethods.Methods
+                interfaceMethods.DeclaredMethods
                     .Where(context => context.UnmanagedToManagedStub.Diagnostics.All(diag => diag.Descriptor.DefaultSeverity != DiagnosticSeverity.Error))
                     .Select(context => context.GenerationContext),
                 vtableLocalName);

--- a/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
@@ -417,7 +417,7 @@ namespace Microsoft.Interop
                         comInterfaceAndMethods.DeclaredMethods
                             .Select(m => m.UnmanagedToManagedStub)
                             .OfType<GeneratedStubCodeContext>()
-                            .Where(context => context.Diagnostics.Length == 0)
+                            .Where(context => context.Diagnostics.Any(diag => diag.Descriptor.DefaultSeverity == DiagnosticSeverity.Error))
                             .Select(context => context.Stub.Node)));
         }
 
@@ -580,7 +580,7 @@ namespace Microsoft.Interop
 
             var vtableSlotAssignments = VirtualMethodPointerStubGenerator.GenerateVirtualMethodTableSlotAssignments(
                 interfaceMethods.Methods
-                    .Where(context => context.UnmanagedToManagedStub.Diagnostics.Length == 0)
+                    .Where(context => context.UnmanagedToManagedStub.Diagnostics.Any(diag => diag.Descriptor.DefaultSeverity == DiagnosticSeverity.Error))
                     .Select(context => context.GenerationContext),
                 vtableLocalName);
 

--- a/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
@@ -417,7 +417,7 @@ namespace Microsoft.Interop
                         comInterfaceAndMethods.DeclaredMethods
                             .Select(m => m.UnmanagedToManagedStub)
                             .OfType<GeneratedStubCodeContext>()
-                            .Where(context => context.Diagnostics.Any(diag => diag.Descriptor.DefaultSeverity == DiagnosticSeverity.Error))
+                            .Where(context => context.Diagnostics.All(diag => diag.Descriptor.DefaultSeverity != DiagnosticSeverity.Error))
                             .Select(context => context.Stub.Node)));
         }
 
@@ -580,7 +580,7 @@ namespace Microsoft.Interop
 
             var vtableSlotAssignments = VirtualMethodPointerStubGenerator.GenerateVirtualMethodTableSlotAssignments(
                 interfaceMethods.Methods
-                    .Where(context => context.UnmanagedToManagedStub.Diagnostics.Any(diag => diag.Descriptor.DefaultSeverity == DiagnosticSeverity.Error))
+                    .Where(context => context.UnmanagedToManagedStub.Diagnostics.All(diag => diag.Descriptor.DefaultSeverity != DiagnosticSeverity.Error))
                     .Select(context => context.GenerationContext),
                 vtableLocalName);
 

--- a/src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs
+++ b/src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs
@@ -355,8 +355,6 @@ namespace ComInterfaceGenerator.Unit.Tests
             {
                 TestCode = source,
                 TestBehaviors = TestBehaviors.SkipGeneratedSourcesCheck,
-                // Our fallback mechanism for invalid code for unmanaged->managed stubs sometimes generates invalid code.
-                CompilerDiagnostics = CompilerDiagnostics.None,
             };
             test.ExpectedDiagnostics.AddRange(expectedDiagnostics);
             await test.RunAsync();

